### PR TITLE
Fix `--filter_{input_type}`

### LINKED ISSUE #287: `--filter_{input_type}` ValueError
## The problem

Encountered this one this morning while testing a workflow where `--filter-T1w` is used at the command-line. This was previously working in `0.7.2`, but no longer working as of `0.8.0`. Ended up getting the following error:

```python
Traceback (most recent call last):
  File "/project/6050199/tkai/shared_venvs/scattr-X3WkDvET-py3.9/bin/scattr", line 6, in <module>
    sys.exit(main())
  File "/scratch/tkai/0_dev/scattr/scattr/run.py", line 19, in main
    app.run_snakemake()
  File "/home/tkai/project/shared_venvs/scattr-X3WkDvET-py3.9/lib/python3.9/site-packages/snakebids/app.py", line 161, in run_snakemake
    update_config(self.config, args)
  File "/home/tkai/project/shared_venvs/scattr-X3WkDvET-py3.9/lib/python3.9/site-packages/snakebids/app.py", line 257, in update_config
    pybids_inputs[input_type]["filters"].update(arg_filter_dict)
ValueError: too many values to unpack (expected 2)
```

using the following call:

```bash
scattr /scratch/tkai/0_sandbox/scattr_test/bids /scratch/tkai/0_sandbox/scattr_test/derivatives participant --slurm_tmpdir --dwi_dir /scratch/tkai/0_sandbox/scattr_test/derivatives/prepdwi_v0.0.13/ --freesurfer_dir /scratch/tkai/0_sandbox/scattr_test/derivatives/fmriprep_v21.0.0_freesurfer/ --fs-license $FS_LICENSE_FILE --sl-count 5000 -p --cores 8 --use-singularity --bzero_thresh 15 --force-output --verbose --filter-T1w desc=synthstrip
```

I don't think we changed how the `update_config` call worked. May just be how I am calling it, but wanted to raise the issue in case it is something bigger.

## Steps to reproduce

* Run a workflow that uses snakebids `0.8.0` with `--filter-{input_type}` flag.

## Screenshots

If you have any helpful screenshots or a file you tried to upload, please include them!

## Environment

* `snakebids` version (if applicable): `0.8.0`
* Operating system (e.g. Ubuntu 20.04, Windows 10): Ubuntu 20.04

### PR DESCRIPTION
**Pull request by kaitj**
## Proposed changes

Resolves #287. Noticed that the `args_dict` only returned the `key` in a list instead of a dict, and originates from changes to `_resolve_path`. The `--filter_{input_type}` calls were read in correctly. I think this bug may have been introduced when we changed some of the `List` types to `Iterable`/`Sequence`. In any case this issue was resolved by changing `Iterable` to `Sequence` in `_resolve_path`.

My :brain: can't come up with a coherent explanation why this bug was happening, but I think it probably has something to do with trying to recursively resolve the "path" in the dict when the condition `isInstance(path_candidate, Iterable)` was met in the original code.

As noted in #287, we should probably add testing for this (#94?). In the meantime, it would be good to get this patched and released to get this functionality working again.

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you are unsure about any of the choices, don't hesitate to ask!

- [x] Changes have been tested to ensure that fix is effective or that a feature works.
- [x] Changes pass the unit tests
- [x] I have included necessary documentation or comments (as necessary)
- [x] Any dependent changes have been merged and published

## Notes
All PRs will undergo the unit testing before being reviewed. You may be requested to explain or make additional changes before the PR is accepted.

### CODE REVIEW

**Review by tkkuehn — APPROVED** [review]
Wow, thanks for tracking this down. I didn't test this myself but assuming you did and it worked I'm happy to merge it

**Comment by tkkuehn** [comment]
Just to nail down precisely what's happening here (for future reference in particular for #94 and for my own edification):

We add an argument to filter the entities for every pybids_input:

https://github.com/akhanf/snakebids/blob/825cf492833d9d2f0c3a90a3630a9d0a92321df3/snakebids/cli.py#L213-L223

We define a `KeyValue` action that parses the passed filters, which sets `args.filter_{input_type}` to a dict where every key is the entity to filter, and every value is the filtered value of that entity.

https://github.com/akhanf/snakebids/blob/825cf492833d9d2f0c3a90a3630a9d0a92321df3/snakebids/cli.py#L24-L43

We call `_resolve_path` on every value of the namespace (including the whole dict assigned to `args.filter_{input_type}`:

https://github.com/akhanf/snakebids/blob/825cf492833d9d2f0c3a90a3630a9d0a92321df3/snakebids/cli.py#L279

`_resolve_path` handles the value if it's a `Sequence` (formerly `Iterable`) or PathLike. Because dictionaries are iterable, and they iterate only over their keys, `_resolve_path` attempted to resolve all the keys in the dict assigned to `args.filter_{input_type}`, discarding the values. Switching to `Sequence` fixes this because dictionaries aren't `Sequence`s.

https://github.com/akhanf/snakebids/blob/825cf492833d9d2f0c3a90a3630a9d0a92321df3/snakebids/cli.py#L318-L343

**Comment by pvandyken** [comment]
Thanks for tracking this down Jason. I agree we don't need a full test suite with this PR, but I would say at least we should add a test case for dictionaries

File: snakebids/cli.py [hunk 1]
```diff
@@ -316,7 +316,7 @@ def _make_underscore_dash_aliases(name: str) -> set[str]:
 
 
 def _resolve_path(
-    path_candidate: Iterable["os.PathLike[str] | str"] | "os.PathLike[str]" | str,
+    path_candidate: Sequence["os.PathLike[str] | str"] | "os.PathLike[str]" | str,
```
[the remark this answers is not in the document]
Comment (pvandyken) [reply snakebids/cli.py:319]: Strings actually don't get converted into paths, and that's by design

[the remark this answers is not in the document]
Comment (pvandyken) [reply snakebids/cli.py:319]: True, yeah, so that would be the first overload, followed by the generic

**Comment by kaitj** [comment]
> Thanks for tracking this down Jason. I agree we don't need a full test suite with this PR, but I would say at least we should add a test case for dictionaries

Yeah, I agree. I was thinking about this a little bit this morning, but hadn't wanted to add it in at that moment without more thought. I'm wondering if we can add something to the existing `TestResolvePath`.

**Comment by tkkuehn** [comment]
Between the two of us, Jason and I weren't able to satisfactorily come up with a type signature for this function. The proposal above (correctly, I think) fails (counterexample: `tuple[tuple[int]] -> list[list[int]]`, not `tuple[list[int]]`). I messed around a bit with recursive sequence `TypeAlias`es, but it seems like as long as we don't constrain the underlying `TypeVar`, pyright is going to complain (I think this gets at some of the issues I'm seeing) -- let me know if you want to see an example of what I'm talking about.

Anyway I think the behaviour of this function is pretty straightforward (famous last words...) and it only gets called in one place so I think I'm okay with just leaving it untyped, especially once we get it properly tested.

**Comment by kaitj** [comment]
Having some issues with testing locally so relying on the actions here and may have overlooked / oversimplified the test case.

**Comment by pvandyken** [comment]
> Between the two of us, Jason and I weren't able to satisfactorily come up with a type signature for this function. The proposal above (correctly, I think) fails (counterexample: tuple[tuple[int]] -> list[list[int]], not list[tuple[int]]). I messed around a bit with recursive sequence TypeAliases, but it seems like as long as we don't constrain the underlying TypeVar, pyright is going to complain (I think this gets at some of the issues I'm seeing) -- let me know if you want to see an example of what I'm talking about.

Fair enough, it sounds like we've hit the limits. I agree about not spending time on this... I would do 

```
@overload
def _resolve_path(path_candidate: Sequence[Any]) -> "list[Any]":
    ...

@overload
def _resolve_path(path_candidate: _T) -> _T:
    ...

def _resolve_path(path_candidate: Any) -> Any:
    ...
```

Which is slightly better than completely untyped, while still being accurate

File: snakebids/cli.py [hunk 2]
```diff
@@ -334,10 +331,13 @@ def _resolve_path(
         Otherwise, the argument unchanged.
     """
 
-    if isinstance(path_candidate, Iterable) and not isinstance(path_candidate, str):
-        return [_resolve_path(p) for p in path_candidate]
+    if isinstance(path_candidate, Sequence) and not isinstance(path_candidate, str):
+        return [
+            _resolve_path(p)
+            for p in path_candidate  # type: ignore[reportUnknownVariableType]
+        ]  # type: ignore[reportUnknownArgumentType]
```
Comment (pvandyken) [inline snakebids/cli.py:338]: Pylance wants this moved up to line 336, at least on my machine, but if it passes the checks don't worry

**Review by pvandyken — APPROVED** [review]

File: snakebids/cli.py (hunk 2 above)
Comment (kaitj) [reply snakebids/cli.py:338]: Was the same on my machine for Pylance, I just moved it up and will make the push soon.

**Comment by kaitj** [comment]
Tristan and I had a brief discussion about the overloads while we were trying to resolve the type signatures. In any case I've added it back in, but think this merge should be held off until we have that discussed.

**Comment by pvandyken** [comment]
> Tristan and I had a brief discussion about the overloads while we were trying to resolve the type signatures. In any case I've added it back in, but think this merge should be held off until we have that discussed.

I'm honestly not too concerned about the overloads, it's a one-off private method... If Tristan has a different opinion just go with his

**Comment by tkkuehn** [comment]
I am for the record okay with anything that keeps pyright happy, so I guess my final word is try [Peter's suggestion](https://github.com/akhanf/snakebids/pull/289#issuecomment-1535544033) and if Pyright doesn't like it leave it untyped

**Review by tkkuehn — APPROVED** [review]

**Pull request merged**